BUG: Correct "remez" error message

sirtom67 opened the pull request:
The error case for remez is erroneously reporting the max iterations has been achieved, this is not the case.
The actual number of iterations is less and the failure is due to a decreasing deviation on an early iteration. This is due to the error being too small - smaller than round-off - and indicates specifications that are too lax. A more accurate error message indicates that the number of iterations is less than maximum, and that it might be averted by increasing the transition band width.

Here's a filter that triggers the incorrect error. In fact, the number of iterations is much fewer, something like 4-5.
```
In [29]: remez(542, [0,0.31, .4 ,1],[1.0, 0.0],Hz=2.0);
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-29-7cddd94452b2> in <module>()
----> 1 remez(542, [0,0.31, .4 ,1],[1.0, 0.0],Hz=2.0);

/usr/local/lib/python3.5/dist-packages/scipy/signal/fir_filter_design.py in remez(numtaps, bands, desired, weight, Hz, type, maxiter, grid_density, fs)
    747     bands = np.asarray(bands).copy()
    748     return sigtools._remez(numtaps, bands, desired, weight, tnum, fs,
--> 749                            maxiter, grid_density)
    750 
    751 

ValueError: Failure to converge after 25 iterations.
```

larsoner commented:
AppVeyor failures seem legitimately related to this PR:

https://ci.appveyor.com/project/scipy/scipy/build/1.0.3431/job/xfn7bikfl7t8widg#L5306

Otherwise looks good to me!

larsoner commented on scipy/signal/sigtoolsmodule.c:
```diff
@@ -1268,14 +1269,15 @@ static PyObject *sigtools_remez(PyObject *NPY_UNUSED(dummy), PyObject *args)
     h = (PyArrayObject *)PyArray_SimpleNew(1, &ret_dimens, NPY_DOUBLE);
     if (h == NULL) goto fail;
 
+    int niter = -1;
```
Probably need to move this up to around 1211

sirtom67 commented:
Eric: Hi! Thanks for the review, hopefully it passes this time.
- Tom

larsoner commented:
Do you think it's worth adding a tiny test that you get an error message that reports a reasonable number of iterations, rather than what you were getting on `master`?

sirtom67 commented:
Yes, I think that would be a good addition. I recently ported "remez" to Julia's DSP package - that's how I discovered this error - and added a check for the error to the unittests with this design:
```
@testset "remez_error_no_converge_transition_band_too_wide" begin
    @test_throws ErrorException remez(151, [0, 0.1, 0.4, 0.5], [1.0, 0.0])
end
````
FYI there is also no warning currently when "maxiter" iterations are reached without convergence. I added that warning to the Julia version (sorry for the julia code, just want to be as clear as possible for whoever adds it to the scipy - the C is line-for-line equivalent so the warning belongs right before the "break" statement after label L100):
```
        #
        # Start next iteration
        #        
      @label L100
        iext[nzz] = ngrid + 1
        niter += 1
        if niter > maxiter
            Compat.@warn("remez() iteration count exceeds maxiter = $maxiter, filter is not converged; try increasing maxiter")
            # the filter is returned in its current, unconverged state.
            break
        end
```
and added a check:
```
@testset "remez_warn_no_converge_after_maxiter_iterations" begin
    warn_check(msg) = contains(msg, "filter is not converged")
    @test_warn warn_check remez(180, [0, 0.375, 0.4, 0.5, 0.525, 1.0], [0.0, 1.0, 0.0]; Hz=2.0)
end
```

larsoner commented:
Thanks @sirtom67
